lib/guardian/plug/sliding_cookie.ex

```diff
@@ -0,0 +1,107 @@
+    `sliding_cookie/3` function is not provided, in which case it is called
+    with a type of `:implementation_fault` and reason `:no_sliding_cookie_fn`.
+
+    This, like all other Guardian plugs, requires a Guardian pipeline to be setup.
+    It requires an implementation module, an error handler and a key.
+
+    These can be set either:
+
+    1. Upstream on the connection with `plug Guardian.Pipeline`
+    2. Upstream on the connection with `Guardian.Pipeline.{put_module, put_error_handler, put_key}`
+    3. Inline with an option of `:module`, `:error_handler`, `:key`
+
+    Nothing is done with the token, refreshed or not, no errors are handled as validity and expiry
+    can be checked by the VerifyCookie and EnsureAuthenticated plugs respectively.
+
+    Options:
+
+    * `:key` - The location of the token (default `:default`)
+    * `:sliding_cookie` - The minimum TTL remaining after which a replacement will be issued. Defaults to configured values.
+    * `:halt` - Whether to halt the connection in case of error. Defaults to `true`.
+
+    The `:sliding_cookie` config (or plug option) should be the same format as `:ttl`, for example
+    `{1, :hour}`, and obviously it should be less than the prevailing `:ttl`.
+    """
+
+    import Plug.Conn
+    import Guardian.Plug, only: [find_token_from_cookies: 2, maybe_halt: 2]
+
+    alias Guardian.Plug.Pipeline
+
+    import Guardian, only: [ttl_to_seconds: 1, decode_and_verify: 4, timestamp: 0]
+
+    @behaviour Plug
+
+    @impl Plug
+    @spec init(opts :: Keyword.t()) :: Keyword.t()
+    def init(opts), do: opts
+
+    @impl Plug
+    @spec call(conn :: Plug.Conn.t(), opts :: Keyword.t()) :: Plug.Conn.t()
+    def call(%{req_cookies: %Plug.Conn.Unfetched{}} = conn, opts) do
+      conn
+      |> fetch_cookies()
+      |> call(opts)
+    end
+
+    def call(conn, opts) do
+      with {:ok, token} <- find_token_from_cookies(conn, opts),
+           module <- Pipeline.fetch_module!(conn, opts),
+           {:ok, ttl_softlimit} <- sliding_window(module, opts),
+           {:ok, %{"exp" => exp} = claims} <- decode_and_verify(module, token, %{}, opts),
+           {:ok, resource} <- module.resource_from_claims(claims),
+           true <- timestamp() >= exp - ttl_softlimit,
+           {:ok, new_c} <- module.sliding_cookie(claims, resource, opts) do
+        conn
+        |> Guardian.Plug.remember_me(module, resource, new_c)
+      else
+        {:error, :not_implemented} ->
+          conn
+          |> Pipeline.fetch_error_handler!(opts)
+          |> apply(:auth_error, [conn, {:implementation_fault, :no_sliding_cookie_fn}, opts])
+          |> maybe_halt(opts)
+
+        _ ->
+          conn
+      end
+    end
+
+    defp sliding_window(module, opts) do
+      case Keyword.get(opts, :sliding_cookie, module.config(:sliding_cookie)) do
+        nil ->
+          :no_sliding_window
+
+        ttl_descr ->
+          {:ok, ttl_to_seconds(ttl_descr)}
```

Comment:
Wouldn't it be simpler to expect a number instead of a tuple?

If people want readability on what the unit of the number is they can always use `:timer` like `:timer.hours(2)` 🤷‍♂

Reply:
I don't have much of an opinion as such on that, but I do think it's nice to be consistent and this is how the TTL config/option is specified.